### PR TITLE
Selective ooc muting

### PR DESCRIPTION
**Pull request by Jordie0608**
Adds ability for players to selectively ignore specific players's OOC messages. Messages from admins can't be ignored unless they're deadminned at the time. You can't ignore yourself.

### CODE REVIEW

**Comment by Ricotez** [comment]
:+1: if you default the list to include punnches

**Comment by Jordie0608** [comment]
See branch name

File: code/game/verbs/ooc.dm [hunk 1]
```diff
@@ -138,3 +138,23 @@ var/global/normal_ooc_colour = "#002eb8"
...
+	prefs.ignoring ^= C.key
+	src << "You are [prefs.ignoring.Find(src.key) ? "now" : "no longer"] ignoring [C.key] on the OOC channel."
+	prefs.save_preferences()
+
+/client/verb/select_ignore()
+	set name = "Ignore"
+	set category = "OOC"
+	set desc ="Ignore a player's messages on the OOC channel"
+
+	var/list/keys = list()
```
Comment (RemieRichards) [inline code/game/verbs/ooc.dm]: I think we have a client list.

File: code/game/verbs/ooc.dm [hunk 2]
```diff
@@ -138,3 +138,23 @@ var/global/normal_ooc_colour = "#002eb8"
 		return
 
 	show_note(usr, null, 1)
+
+/client/proc/ignore_key(client)
+	var/client/C = client
+	prefs.ignoring ^= C.key
```
Comment (RemieRichards) [inline code/game/verbs/ooc.dm]: ```^=``` doesn't seem like the right operator here.
1, C.key isn't a list or a number, so this isn't the list operation or the XOR operation
2. If it was a list you'd be setting A equal to just the unique elements of B, which really makes no sense as just a ```|=``` and a ```-=``` in an if/else would do it.

you're

File: code/game/verbs/ooc.dm (hunk 1 above)
Comment (GinjaNinja32) [inline code/game/verbs/ooc.dm]: `for(var/client/C)` works too, you don't even need a list.

Comment (RemieRichards) [inline code/game/verbs/ooc.dm]: @GinjaNinja32 That's ```in world``` and we don't use that as a list is faster.

Comment (GinjaNinja32) [inline code/game/verbs/ooc.dm]: `for(var/client/C in world)` does *not* work IIRC. In world is faster for base types anyway.

Comment (RemieRichards) [inline code/game/verbs/ooc.dm]: No argument defaults to inworld

```
for loop proc
list
Format:
for (Var [as Type] [in List]) Statement
Args:
Var: A variable to sequentially contain each member of the list.
List: The list to loop through. This defaults to the whole world.
```

**Comment by Robertha89** [comment]
This is the greatest thing i've seen.

File: code/game/verbs/ooc.dm (hunk 1 above)
Comment (GinjaNinja32) [inline code/game/verbs/ooc.dm]: if that's fromm the ref, you should know how accurate it generally is :P

Comment (RemieRichards) [inline code/game/verbs/ooc.dm]: It's from the ref and the bluebook, and the redbook doesn't mention it as erroneous.
Not to mention every other var you use defaults to in world, I very much doubt client would have it's own snowflake code in the base for list proc.

Comment (GinjaNinja32) [inline code/game/verbs/ooc.dm]: Datum doesn't. `for(var/datum/d in world)` won't execute its body, `for(var/datum/d)` will. IIRC, client is the same.

Comment (RemieRichards) [inline code/game/verbs/ooc.dm]: that's very dodgy.

**Comment by Aranclanos** [comment]
make a ooc ban instead

**Comment by RemieRichards** [comment]
No, make an ooc ban ALSO. (iirc admin ooc mutes only last for that round, and the player can speak again next round, so OOC ban is something new and potentially needed)

File: code/game/verbs/ooc.dm [hunk 3]
```diff
@@ -55,7 +55,7 @@
 						C << "<span class='adminobserverooc'><span class='prefix'>OOC:</span> <EM>[keyname][holder.fakekey ? "/([holder.fakekey])" : ""]:</EM> <span class='message'>[msg]</span></span>"
 				else
 					C << "<font color='[normal_ooc_colour]'><span class='ooc'><span class='prefix'>OOC:</span> <EM>[holder.fakekey ? holder.fakekey : key]:</EM> <span class='message'>[msg]</span></span></font>"
-			else
+			else if(!C.prefs.ignoring.Find(key))
 				C << "<font color='[normal_ooc_colour]'><span class='ooc'><span class='prefix'>OOC:</span> <EM>[keyname]:</EM> <span class='message'>[msg]</span></span></font>"
```
Comment (RemieRichards) [inline code/game/verbs/ooc.dm]: Can you maybe show that they said something, to avoid confusion? eg: "That dick you hate: Muted in your preferences"

**Comment by Bawhoppen** [comment]
merge quickly pls

File: code/game/verbs/ooc.dm (hunk 1 above)
Comment (Jordie0608) [inline code/game/verbs/ooc.dm]: There's a client list, will fix.

File: code/game/verbs/ooc.dm (hunk 2 above)
Comment (Jordie0608) [inline code/game/verbs/ooc.dm]: Yeah that should be an OR, seems I used the wrong operator

File: code/game/verbs/ooc.dm (hunk 3 above)
Comment (Jordie0608) [inline code/game/verbs/ooc.dm]: That'd partly defeat the point of ignoring them and while I can't speak for everyone I'd find that to be be rather annoying to see, especially if the person chattered a lot.

**Comment by Jordie0608** [comment]
OOC ban I'll make a different PR for.

File: code/game/verbs/ooc.dm (hunk 3 above)
Comment (RemieRichards) [inline code/game/verbs/ooc.dm]: I just don't want people to be confused about speaking in say, a 3-way conversation where they literally can't tell one of the people is speaking, but w/e.

Comment (Jordie0608) [inline code/game/verbs/ooc.dm]: It works in IRC

Comment (RemieRichards) [inline code/game/verbs/ooc.dm]: IRC is not a perfect system.

**Comment by Fox-McCloud** [comment]
All the yes :+1:

**Comment by Memendia** [comment]
I probably newroundsexy.ogg'd hippie by porting this lmao

**Comment by freerealestate** [comment]
Great idea, though it's a shame you can't mute admins
CsEtRiTcAkIyN PEOPLE use the coloured text and the respect that comes from the authority to make it their personal blog
It'd be better if they could see who had blocked them so they could PM them if they needed to talk to them about their conduct in OOC

**Comment by optimumtact** [comment]
What do you mean you can't mute admins?

**Comment by Bawhoppen** [comment]
I feel like if an admin has something important to say, they should make an adminnotice OR ahelp it specifically to you if its private matter

**Comment by optimumtact** [comment]
Yah I think you should be able to block an admin from your ooc

**Comment by PKPenguin321** [comment]
this is long overdue :+1:

**Comment by joonoro** [comment]
:+1: 

I was going to add it myself but looks like you beat me to it, nice job.

MERGE *NOW*.

**Comment by SweaterKittens** [comment]
Sweet jesus, thank you for this.

**Comment by MrPerson** [comment]
Might be wise to ask the admins if they want to be mutable or not. I say leave it up to them

**Comment by freerealestate** [comment]
Why would they want that

**Comment by optimumtact** [comment]
Why should an admin be allowed to harrass you in ooc but nobody else can? They can just use adminhelp or narrates.

**Comment by Shadowlight213** [comment]
@RemieRichards @Razharas @Cheridan @Miauw62 @Aranclanos @phil235 

Merge please.

**Comment by freerealestate** [comment]
#12652

**Pull request merged**